### PR TITLE
Clean up SSL certificate based on domain

### PR DESCRIPTION
**Pull request by zhenghuiwang**
Fixes https://github.com/kubeflow/kubeflow/issues/4227
Fixes https://github.com/kubeflow/kfctl/issues/42

- Fetch the SSL certificate domain info from Compute Beta API.
- Wait for deletion operations of forwarding rules and targetHttpsProxy to finish.
- Fix the wait logic of deleting deployment: current logic is to wait operation one by one; this PR changes it to wait once for all pending ops.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/testing/486)

### CODE REVIEW

**Comment by zhenghuiwang** [comment]
/assign @jlewi 

Ready for your review!

File: py/kubeflow/testing/cleanup_ci.py [hunk 1]
```diff
@@ -579,10 +583,27 @@ def cleanup_health_checks(args):
   logging.info("Matched health checks:\n%s", "\n".join(matched))
   logging.info("Finished cleanup firewall rules")
 
+def get_ssl_certificate_domain(certificate):
```
Comment (jlewi) [inline py/kubeflow/testing/cleanup_ci.py]: Is this not available via the API?
If we can't get it via the API then we could shell out to gcloud 

```
gcloud beta --project=kubeflow-ci-deployment compute ssl-certificates list --format=yaml
```

and just parse the result

File: py/kubeflow/testing/cleanup_ci.py [hunk 2]
```diff
@@ -787,6 +788,23 @@ def execute_rpc(rpc):
   """Execute a Google RPC request with retries."""
   return rpc.execute()
 
+# Wait for 'ops' to finish in 'max_wait_mins' or return the remaining ops.
+# operation_resource must implement 'get()' method.
+def wait_ops_max_mins(operation_resource, args, ops, max_wait_mins=15):
```
Comment (jlewi) [inline py/kubeflow/testing/cleanup_ci.py]: What was the bug in the existing implementation?

File: py/kubeflow/testing/cleanup_ci.py (hunk 1 above)
Comment (jlewi) [reply py/kubeflow/testing/cleanup_ci.py]: Oh I see it looks we are now getting the domain from the certificate in the managed domains field

File: py/kubeflow/testing/cleanup_ci.py [hunk 3]
```diff
@@ -579,10 +583,27 @@ def cleanup_health_checks(args):
   logging.info("Matched health checks:\n%s", "\n".join(matched))
   logging.info("Finished cleanup firewall rules")
 
+def get_ssl_certificate_domain(certificate):
+  if "managed" in certificate and "domains" in certificate["managed"]:
+    # We use one domain per certificate.
+    return certificate["managed"]["domains"][0]
+
+  if not "certificate" in certificate:
```
Comment (jlewi) [inline py/kubeflow/testing/cleanup_ci.py]: Do we need this code path anymore? i.e.e do we need to parse the raw certificates anymore now that we et the domain from the API

Comment (zhenghuiwang) [reply py/kubeflow/testing/cleanup_ci.py]: There are two types of responses. I think I can change the logic to use "subjectAlternativeNames"
```
    {
      "id": "5281947711917522957",
      "creationTimestamp": "2019-10-15T05:43:14.206-07:00",
      "name": "k8s-ssl-fce70c0d80098ce1-daf4c582aeab0540--eace8224f03fad2f",
      "selfLink": "https://www.googleapis.com/compute/beta/projects/kubeflow-ci-deployment/global/sslCertificates/k8s-ssl-fce70c0d80098ce1-daf4c582aeab0540--eace8224f03fad2f",
      "certificate": "-----BEGIN CERTIFICATE-----\nMIIFpTCCBI2gAwIBAgISAwIHO1Jo775Os7kBIdnQ2k3xMA0GCSqGSIb3DQEBCwUA\nMEoxCzAJBgNVBAYTAlVTMRYwFAYDVQQKEw1MZXQncyBFbmNyeXB0MSMwIQYDVQQD\nExpMZXQncyBFbmNyeXB0IEF1dGhvcml0eSBYMzAeFw0xOTEwMTUxMTQyNDVaFw0y\nMDAxMTMxMTQyNDVaMEIxQDA+BgNVBAMTN2tmLXYwLTYtbjAwLmVuZHBvaW50cy5r\ndWJlZmxvdy1jaS1kZXBsb3ltZW50LmNsb3VkLmdvb2cwggEiMA0GCSqGSIb3DQEB\nAQUAA4IBDwAwggEKAoIBAQDy/HU4IxUp2HRjkF2Q0qrEGW7PwADAoL3Qcs/bbz89\nx+/j2Co1fPF7oKcL0+oh5NRei+wTHTydFXxwugIo+9ea53TP6kJhuCUU4zLH89T/\nEvD+YQosw5DfvyTFu0JCrZu19GHKatChmvKxUt7SQFtcAteQ9Jt+uACFmM5SJNP2\nBurowabjkmV4E8on4g0AxfwG7vvacwi/PCNrF+1F4+aGZhUUT1p60XP54/n4GWCP\nIsz0vhr99gVL7Px2X0ARZo7Vx4qarOowxv/FCQkBk0vkdDI/oTdx7sKI4WjUtlC7\nj81F46o/fRCRNV6v0uoUZRzDIXaTyzzGg+tnXzRNluLpAgMBAAGjggKLMIIChzAO\nBgNVHQ8BAf8EBAMCBaAwHQYDVR0lBBYwFAYIKwYBBQUHAwEGCCsGAQUFBwMCMAwG\nA1UdEwEB/wQCMAAwHQYDVR0OBBYEFNcuqjEu3dt3vylEIzMeYkLCxf+MMB8GA1Ud\nIwQYMBaAFKhKamMEfd265tE5t6ZFZe/zqOyhMG8GCCsGAQUFBwEBBGMwYTAuBggr\nBgEFBQcwAYYiaHR0cDovL29jc3AuaW50LXgzLmxldHNlbmNyeXB0Lm9yZzAvBggr\nBgEFBQcwAoYjaHR0cDovL2NlcnQuaW50LXgzLmxldHNlbmNyeXB0Lm9yZy8wQgYD\nVR0RBDswOYI3a2YtdjAtNi1uMDAuZW5kcG9pbnRzLmt1YmVmbG93LWNpLWRlcGxv\neW1lbnQuY2xvdWQuZ29vZzBMBgNVHSAERTBDMAgGBmeBDAECATA3BgsrBgEEAYLf\nEwEBATAoMCYGCCsGAQUFBwIBFhpodHRwOi8vY3BzLmxldHNlbmNyeXB0Lm9yZzCC\nAQMGCisGAQQB1nkCBAIEgfQEgfEA7wB2AF6nc/nfVsDntTZIfdBJ4DJ6kZoMhKES\nEoQYdZaBcUVYAAABbc9xM2gAAAQDAEcwRQIhAOBM0Dc5PS2wU/rU0ED1NFn+5xqo\nWnnuX0rgunUtG3o3AiBK3sgmVZo8KrNPoIzLaqerMOnmFhkZbEfNvy0+EZF5SAB1\nALIeBcyLos2KIE6HZvkruYolIGdr2vpw57JJUy3vi5BeAAABbc9xM2sAAAQDAEYw\nRAIgGPv/oaP1R/svnm5eeEfPnHInlCM455osJNFkMvaTwBsCIGYvT5NjLybJc9OC\nbS6kieLI3VfKK01syZBh107gyN8uMA0GCSqGSIb3DQEBCwUAA4IBAQBxH/u0335n\ns7qKb6ymbc/JwfhZNMlFGdskOZBzA4KPxEzt/v4UgbrwBXmnEhNvED+Y8KIdwSVh\n3rmm9eUzd2DrVkA9iwJk65Gg02JVY4s+X/FHWnLgWrBZurh85VCeCGo+x8RWxcis\nDWs8yVYVesIQmqLAf4/TVOR6/1dP1Vj5PKsTfFdf5k6oWOT1Qc1hkUX71JPrxnn6\nsOuLYKg8Wv6ETHi5iel/sJLjuaMuYectoyH/pfWsCodaXPWSIDO/iKqVlze42gy/\ngB/V5mXoHC5kYJat5w0K5dqOyrPZo9ggNv9GMyCVj15KHemq5j0SkbVKI6Lm4jCE\nNBYfVe5kV4oP\n-----END CERTIFICATE-----\n-----BEGIN CERTIFICATE-----\nMIIEkjCCA3qgAwIBAgIQCgFBQgAAAVOFc2oLheynCDANBgkqhkiG9w0BAQsFADA/\nMSQwIgYDVQQKExtEaWdpdGFsIFNpZ25hdHVyZSBUcnVzdCBDby4xFzAVBgNVBAMT\nDkRTVCBSb290IENBIFgzMB4XDTE2MDMxNzE2NDA0NloXDTIxMDMxNzE2NDA0Nlow\nSjELMAkGA1UEBhMCVVMxFjAUBgNVBAoTDUxldCdzIEVuY3J5cHQxIzAhBgNVBAMT\nGkxldCdzIEVuY3J5cHQgQXV0aG9yaXR5IFgzMIIBIjANBgkqhkiG9w0BAQEFAAOC\nAQ8AMIIBCgKCAQEAnNMM8FrlLke3cl03g7NoYzDq1zUmGSXhvb418XCSL7e4S0EF\nq6meNQhY7LEqxGiHC6PjdeTm86dicbp5gWAf15Gan/PQeGdxyGkOlZHP/uaZ6WA8\nSMx+yk13EiSdRxta67nsHjcAHJyse6cF6s5K671B5TaYucv9bTyWaN8jKkKQDIZ0\nZ8h/pZq4UmEUEz9l6YKHy9v6Dlb2honzhT+Xhq+w3Brvaw2VFn3EK6BlspkENnWA\na6xK8xuQSXgvopZPKiAlKQTGdMDQMc2PMTiVFrqoM7hD8bEfwzB/onkxEz0tNvjj\n/PIzark5McWvxI0NHWQWM6r6hCm21AvA2H3DkwIDAQABo4IBfTCCAXkwEgYDVR0T\nAQH/BAgwBgEB/wIBADAOBgNVHQ8BAf8EBAMCAYYwfwYIKwYBBQUHAQEEczBxMDIG\nCCsGAQUFBzABhiZodHRwOi8vaXNyZy50cnVzdGlkLm9jc3AuaWRlbnRydXN0LmNv\nbTA7BggrBgEFBQcwAoYvaHR0cDovL2FwcHMuaWRlbnRydXN0LmNvbS9yb290cy9k\nc3Ryb290Y2F4My5wN2MwHwYDVR0jBBgwFoAUxKexpHsscfrb4UuQdf/EFWCFiRAw\nVAYDVR0gBE0wSzAIBgZngQwBAgEwPwYLKwYBBAGC3xMBAQEwMDAuBggrBgEFBQcC\nARYiaHR0cDovL2Nwcy5yb290LXgxLmxldHNlbmNyeXB0Lm9yZzA8BgNVHR8ENTAz\nMDGgL6AthitodHRwOi8vY3JsLmlkZW50cnVzdC5jb20vRFNUUk9PVENBWDNDUkwu\nY3JsMB0GA1UdDgQWBBSoSmpjBH3duubRObemRWXv86jsoTANBgkqhkiG9w0BAQsF\nAAOCAQEA3TPXEfNjWDjdGBX7CVW+dla5cEilaUcne8IkCJLxWh9KEik3JHRRHGJo\nuM2VcGfl96S8TihRzZvoroed6ti6WqEBmtzw3Wodatg+VyOeph4EYpr/1wXKtx8/\nwApIvJSwtmVi4MFU5aMqrSDE6ea73Mj2tcMyo5jMd6jmeWUHK8so/joWUoHOUgwu\nX4Po1QYz+3dszkDqMp4fklxBwXRsW10KXzPMTZ+sOPAveyxindmjkW8lGy+QsRlG\nPfZ+G6Z6h7mjem0Y+iWlkYcV4PIWL1iwBi8saCbGS5jN2p8M+X+Q7UNKEkROb3N6\nKOqkqm57TH2H3eDJAkSnh6/DNFu0Qg==\n-----END CERTIFICATE-----\n",
      "subjectAlternativeNames": [
        "kf-v0-6-n00.endpoints.kubeflow-ci-deployment.cloud.goog"
      ],
      "expireTime": "2020-01-13T03:42:45.000-08:00",
      "kind": "compute#sslCertificate"
    },
 ```
and
```
    {
      "id": "2503326492445235298",
      "creationTimestamp": "2019-10-17T06:03:09.747-07:00",
      "name": "mcrt-13a09bbd-e720-471b-bbc1-6ecfaf0df0f8",
      "selfLink": "https://www.googleapis.com/compute/beta/projects/kubeflow-ci-deployment/global/sslCertificates/mcrt-13a09bbd-e720-471b-bbc1-6ecfaf0df0f8",
      "managed": {
        "domains": [
          "kfctl-ba61.endpoints.kubeflow-ci-deployment.cloud.goog"
        ],
        "status": "PROVISIONING",
        "domainStatus": {
          "kfctl-ba61.endpoints.kubeflow-ci-deployment.cloud.goog": "FAILED_NOT_VISIBLE"
        }
      },
      "type": "MANAGED",
      "kind": "compute#sslCertificate"
    },
```

File: py/kubeflow/testing/cleanup_ci.py (hunk 2 above)
Comment (zhenghuiwang) [reply py/kubeflow/testing/cleanup_ci.py]: The bug is that the existing wait logic is inside the for-loop of deployments. Hence there is always one op in `delete_ops`, i.e. the function wait for at least 30secs for each deletion.

I think the intention is to kick off all deletion ops and then wait for them to complete. So in the code below i move the `delete_ops` and wait logic out of the deployment for-loop.

**Comment by zhenghuiwang** [comment]
@jlewi thanks for the comments! I updated the PR.

**Pull request merged**